chunks.py

```diff
@@ -0,0 +1,130 @@
+import asyncio
+import io
+import math
+import os
+from concurrent.futures import ThreadPoolExecutor
+
+import config
+from progresslogger import Progress
+
+
+def calculate_number_of_chunks(file_size, chunk_size):
+    return math.ceil(int(file_size) / chunk_size)
+
+
+class Chunks:
+    def __init__(self, file_downloader, file_name, file_size, *chunks: 'Chunk'):
+        def schedule(chunk):
+            return loop.run_in_executor(executor, chunk.download, file_downloader)
+
+        async def write_files():
+            async def await_chunks():
+                for chunk in chunks:
+                    await chunk
+
+            def write_final_file_from_partial_files():
+                with open(file_name, "wb") as final_file:
+                    for chunk in chunks:
+                        with open(chunk.file_name, "rb") as partial_file:
+                            final_file.write(partial_file.read())
+
+            def remove_partial_files():
+                for chunk in chunks:
+                    os.remove(chunk.file_name)
+
+            try:
+                await await_chunks()
+                write_final_file_from_partial_files()
+            finally:
+                remove_partial_files()
+
+        self.file_size = file_size
+        self.chunks = chunks
+        loop = asyncio.get_event_loop()
+        executor = ThreadPoolExecutor(max_workers=5)
+        self.download_chunks_tasks = tuple(map(schedule, chunks))
+        self.create_final_file_task = loop.create_task(write_files())
+
+    def __await__(self):
+        yield from asyncio.gather(self.create_final_file_task)
+
+    def __len__(self):
+        return len(self.chunks)
+
+    def __str__(self):
+        return str(tuple(map(str, self.chunks)))
+
+    def __iter__(self):
+        yield from self.chunks
+
+    def __getitem__(self, item):
+        return self.chunks[item]
+
+    @staticmethod
+    def create_from(file_downloader, file_name, file_size, chunk_size) -> 'Chunks':
+        number_of_chunks = calculate_number_of_chunks(file_size, chunk_size)
+        chunk_range = range(number_of_chunks)
+        # print(number_of_chunks, list(chunk_range))
```

Comment:
You can either use the `logger.d` or remove this line

Reply:
Removed.